### PR TITLE
add Block data-structure&&move TransactionBase to Transaction && add encode and decode interfaces into BlockHeader and Transaction

### PR DESCRIPTION
**Pull request by cyjseagull**
1. move TransactionBase to Transaction && add encode and decode interfaces into BlockHeader and Transaction to hide rlp-decode-encode related logicals && modify unit test correspondingly after interface updated
2. add Block data-structure
3. add BlockFieldSize to represent total fields of block data structure

### CODE REVIEW

File: libethcore/Common.cpp [hunk 1]
```diff
@@ -39,6 +39,7 @@ const unsigned c_protocolVersion = 63;
 const unsigned c_minorProtocolVersion = 2;
 const unsigned c_databaseBaseVersion = 9;
 const unsigned c_databaseVersionModifier = 0;
+const unsigned int BlockFieldSize = 4;
```
Comment (wheatli) [inline libethcore/Common.cpp]: set c_ as prefix for constant var

File: libethcore/Block.h [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    Block(Block const& _block);
+    /// assignment operator
+    Block& operator=(Block const& _block);
+
+    ///-----opearator overloads of Block
+    /// operator ==
+    bool operator==(Block const& _block) const
+    {
+        return m_blockHeader == _block.blockHeader() && m_headerHash == _block.headerHash() &&
+               m_sigList == _block.sigList() && m_transactions == _block.transactions() &&
```
Comment (wheatli) [inline libethcore/Block.h]: the equal of sigList is not a essential condition to check equality of block, one block can be equal to another even if their sigList are not same

Comment (wheatli) [reply libethcore/Block.h]: maybe just need to check the equality of header hash?

File: libethcore/Block.h [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+               m_transactionHashSet == _block.transactionHashSet();
+    }
+    /// operator !=
+    bool operator!=(Block const& _cmp) const { return !operator==(_cmp); }
+    explicit operator bool() const { return bool(m_blockHeader) && (m_transactions.size() != 0); }
```
Comment (wheatli) [inline libethcore/Block.h]: a block is valid even without any tx

File: libethcore/Block.h [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+    void encode(bytes& _out, std::vector<std::pair<u256, Signature>>& sig_list);
+    void encode(bytes& _out)
+    {
+        std::vector<std::pair<u256, Signature>> sig_list;
+        encode(_out, sig_list);
+    }
+
+    ///-----decode functions
+    void decode(bytesConstRef _block);
+
```
Comment (wheatli) [inline libethcore/Block.h]: I thought we may need a validate method to check the validity of block decoded from bytes, maybe it also can be a static method

File: libethcore/Block.h (hunk 2 above)
Comment (cyjseagull) [reply libethcore/Block.h]: I think the equality of Block should depend all fields~

File: libethcore/Block.h (hunk 4 above)
Comment (cyjseagull) [reply libethcore/Block.h]: RLP block_rlp = BlockHeader::extractBlock(_block_bytes); 
is the static function of BlockHeader, this function check the validity of the whole block.

decode function callback this function to ensure the validity of the block
if the block is invalid, it'll throw exceptions.

below is the definition of extractBlock:
```c++
RLP BlockHeader::extractBlock(bytesConstRef _block)
{
    RLP root(_block);
    if (!root.isList())
        BOOST_THROW_EXCEPTION(InvalidBlockFormat() << errinfo_comment("Block must be a list")
                                                   << BadFieldError(0, _block.toString()));
    // check block header(must be list)
    if (!root[0].isList())
        BOOST_THROW_EXCEPTION(InvalidBlockFormat() << errinfo_comment("Block header must be a list")
                                                   << BadFieldError(0, root[0].data().toString()));
    // check the transactions(must be list)
    if (!root[1].isList())
        BOOST_THROW_EXCEPTION(InvalidBlockFormat()
                              << errinfo_comment("Block transactions must be a list")
                              << BadFieldError(1, root[1].data().toString()));
    return root;
}

File: libethcore/Common.cpp (hunk 1 above)
Comment (cyjseagull) [reply libethcore/Common.cpp]: already modified, please refer to https://github.com/FISCO-BCOS/lab-bcos/pull/66/files#diff-4e01855c4603e806caac0a73498394caR42

File: libethcore/Block.h (hunk 3 above)
Comment (cyjseagull) [reply libethcore/Block.h]: already modified to `bool(m_blockHeader)` , please refer to: https://github.com/FISCO-BCOS/lab-bcos/pull/66/files#diff-617647667e5d18f6fb4bf70296a1e638R54

File: libethcore/Block.h [hunk 5]
```diff
@@ -47,70 +47,78 @@ class Block
...
+    {
+        m_blockHeader = _blockHeader;
+        noteBlockChange();
     }
-    void setBlockHeader(BlockHeader const& _blockHeader) { m_blockHeader = _blockHeader; }
+    /// set sig list
+    void inline setSigList(std::vector<std::pair<u256, Signature>> const& _sigList)
+    {
+        m_sigList = _sigList;
+        noteBlockChange();
```
Comment (wheatli) [inline libethcore/Block.h]: sigList will not influence block_hash, don't notBlockChange here

File: libethcore/Block.cpp [hunk 6]
```diff
@@ -177,25 +152,35 @@ void Block::decode(bytesConstRef _block_bytes)
...
-    m_headerHash = h256(block_rlp[2].data());
+    m_headerHash = h256(block_rlp[2].toInt<u256>());
     /// get sig_list
     m_sigList = block_rlp[3].toVector<std::pair<u256, Signature>>();
     noteChange();
+    noteBlockChange();
+}
+
+/// get block hash
+h256 Block::hash()
```
Comment (wheatli) [inline libethcore/Block.cpp]: I think it's the same as headerHash()

File: libethcore/Block.h [hunk 7]
```diff
@@ -121,13 +129,14 @@ class Block
...
-    h256Hash m_transactionHashSet;
     /// m_transactions converted bytes, when m_transactions changed,
     /// should refresh this catch when encode
     bytes m_txsCache;
-    /// means whether m_transactions has been changed
-    bool m_transaction_changed = false;
+    h256 m_blockHash;
+
+    /// mutex for block hash
+    mutable Mutex m_hashLock;
```
Comment (wheatli) [inline libethcore/Block.h]: The multithread access control is not essential part for a pure data carrier struct

File: libethcore/Block.cpp [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+ * @param _out : bytes of generated block
+ * @param block_header : bytes of the block_header
+ * @param hash : hash of the block hash
+ * @param sig_list : signature list
+ */
+void Block::encode(bytes& _out, bytesConstRef block_header, h256 const& hash,
+    std::vector<std::pair<u256, Signature>>& sig_list)
+{
+    /// refresh transaction list cache
+    bytes txsCache = encodeTransactions();
```
Comment (wheatli) [inline libethcore/Block.cpp]: there is a reduntant memory copy of txsCache,  make the encodeTransaction return const& is a method to omit the copy

**Pull request merged**